### PR TITLE
[RNMobile] Add basic upload functionality to VideoPress block

### PR DESCRIPTION
**Pull request by fluiddot**
Part of https://github.com/wordpress-mobile/gutenberg-mobile/issues/5557.

## Proposed changes:

- Implement the native version of the following components:
   - `UploaderProgress`
   - `VideoPressUploader`
- Enable basic upload functionality in the block.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

N/A

## Does this pull request change what data or activity we track or use?

N/A

## Testing instructions:

**Preparation:**
1. Create a post.
2. Add VideoPress block.
3. Observe that a placeholder is shown in the block.
4. Observe that the upload options are shown automatically.

### Upload local video
1. Select the "Choose from device" option.
2. Select a video.
3. Observe that a progress bar is displayed while the upload is in progress.
4. Once the upload finishes, observe that the placeholder is no longer displayed.
5. Preview the post.
6. Observe that the block is displayed and that the video can be played.

### Take a video

1. Select the "Take a Video" option.
2. Record a video and upload it.
3. Observe that a progress bar is displayed while the upload is in progress.
4. Once the upload finishes, observe that the placeholder is no longer displayed.
5. Preview the post.
6. Observe that the block is displayed and that the video can be played.

### WordPress Media Library

⚠️ This option is not supported yet. It will be ready once we pass the VideoPress GUID from the main apps when selecting a media item.

#### Valid URL
1. Select the "Insert from URL" option.
2. Add a VideoPress URL (e.g. https://videopress.com/v/YzEZpv1G).
3. Observe that the placeholder is no longer displayed.
4. Preview the post.
5. Observe that the block is displayed and that the video can be played.

#### Invalid URL
1. Select the "Insert from URL" option.
2. Add an invalid VideoPress URL (e.g. https://wordpress.com).
3. Observe that an error notice is displayed informing about the invalid URL.
4. Observe that the placeholder is still displayed.

#### Other Apps _(only available on iOS)_
1. Select the "Other Apps" option.
2. Select a video.
3. Observe that a progress bar is displayed while the upload is in progress.
4. Once the upload finishes, observe that the placeholder is no longer displayed.
5. Preview the post.
6. Observe that the block is displayed and that the video can be played.

### Upload fail
1. Select the "Choose from device" option.
2. Select a video.
3. Observe that a progress bar is displayed while the upload is in progress.
4. Turn off internet connection and wait.
5. Observe that a retry message is displayed in the block.

**⚠️ NOTE: The retry logic will be incorporated in another PR ([reference](https://github.com/Automattic/jetpack/pull/29461#discussion_r1138598730)).**

### CODE REVIEW

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/style.native.scss
Comment (SiobhyB) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/style.native.scss:4]: Super small detail... The indentation of lines two and three looks a bit wrong here. If I save this file locally, prettifier corrects the formatting so that they're indented in the same way as the other lines.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (jhnstn) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js:39]: I noticed in the web version the upload is limited to the [first file if multiple are passed in](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.js#L165). Do we need something similar here to limit the uploads to 1 ?

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (SiobhyB) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js]: I noticed that the designs for the native block have a more succinct description than the web i.e. `Add Video` vs. `Embed a video from your media library or upload a new one with VideoPress.` 

I'm not sure if we need the lengthier description on native, given the list of available options pop up as soon as the block's added. The shorter `Add Video` is more inline with descriptions in the other native media blocks, too. I could also see a case for having a lengthier description, though. 

This isn't a blocker by any means, so I'll defer to you on which description you'd prefer to go with. 

| Design  | Actual |
| ------------- | ------------- |
|  |  |

**Review by SiobhyB — DISMISSED** [review]
@fluiddot, whoot, it works great for me! :) 🎉 I went through each of the test cases you outlined for both platforms. I also did some general testing with saving the posts across both the web and native to see if I could break anything, but I couldn't. 

| Test | iOS | Android |
| :---         |     :---:      |          ---: |
| Upload local video   | ✅ | ✅ |
| Take a video     | ✅ | ✅ |
| Insert from URL (valid)     |  ✅      | ✅ |
| Insert from URL (invalid)     |  ✅       | ✅ |
| Other apps    |  ✅       | N/A |

I added two really small comments for your consideration, but nothing major. If you do make changes, I'd be happy to re-review and approve again.

_Edit: Note, I published my review before seeing [Jason's comment](https://github.com/Automattic/jetpack/pull/29461#discussion_r1135983122)._

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js]: Good catch. I copied the placeholder labels from the web version but I agree makes more sense to stick with the native designs. I'll update it to reflect this, thanks 🙇 !

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/style.native.scss
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/style.native.scss:4]: Oh, true, not sure why the indentation changed here. I'll update it 💨 .

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js:39]: We only support uploading one file at a time, so initially, we don't need to check if multiple files are passed. I did a quick check and to enable multiple files, we'd need to pass the [prop `multiple` in the `MediaPlaceholder` component](https://github.com/WordPress/gutenberg/blob/a2b1c1147b9bb8e53a96908625476ab9e5405be2/packages/block-editor/src/components/media-placeholder/index.native.js#L50:L50).

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/style.native.scss
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/style.native.scss:4]: Update applied in https://github.com/Automattic/jetpack/pull/29461/commits/bd5d40ab49b0ea40898ee2b64a28b16c35691409.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js]: I applied this suggestion in https://github.com/Automattic/jetpack/pull/29461/commits/c1e9f00c6814a8af8cbfd3720801418a8f61f612.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (SiobhyB) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js]: @fluiddot, it looks like the `description` import just needs to be removed and then I think all will be good.

**Review by SiobhyB — CHANGES_REQUESTED** [review]

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js]: Ah, my bad 🤦. I've just fixed this in https://github.com/Automattic/jetpack/pull/29461/commits/e720ce221d611535d58441f4d9df8e92e5d32e76. Thanks!

**Review by SiobhyB — DISMISSED** [review]
Approved via the review [here](https://github.com/Automattic/jetpack/pull/29461#pullrequestreview-1339862364) and confirming the recent changes look good. Thanks again @fluiddot! 🙌

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/index.native.js:39]: Cool, just making sure. I too noticed I was only able to select one video at a time.

**Review by jhnstn — DISMISSED** [review]
Just finished testing and everything looks good, nice job!

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
Comment (jhnstn) [inline projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:15]: Non-blocking but curious, why not use `./components/videopress-uploader` for the import path. Shouldn't it resolve to the native version?

Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:15]: I could use `./components/videopress-uploader` as it imports the native version, but TypeScript was complaining in the editor due to missing props. In this case, since the component has different props for each platform, TypeScript by default infers the prop types from the default version of the file (i.e. the web version).

Screenshot 2023-03-15 at 09 46 15

**Comment by fluiddot** [comment]
Not sure why but `E2E Tests / Jetpack connection e2e tests` job is failing, the changes introduced in this PR shouldn't affect the logic that is tested by the job 🤔. I retried a couple of times but it still fails. However, I noticed that the same job is also failing on other opened PRs, so I understand this is a common problem not specific to this PR. Being this said, I'll set this PR as `[Status] Needs Review`.

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:15]: Thanks! I ran into that too late in the day yesterday. Seems to work fine without the explicit path but my editor was complaining.

**Comment by jhnstn** [comment]
> Not sure why but `E2E Tests / Jetpack connection e2e tests` job is failing, the changes introduced in this PR shouldn't affect the logic that is tested by the job 🤔. I retried a couple of times but it still fails. However, I noticed that the same job is also failing on other opened PRs, so I understand this is a common problem not specific to this PR. Being this said, I'll set this PR as `[Status] Needs Review`.

Looks like it's failing on trunk too https://github.com/Automattic/jetpack/actions/runs/4426853630/jobs/7763687984

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:15]: Yep, that's it. In my case using VSCode, it's there where I see the error. Using any of both imports doesn't seem to affect the logic.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.native.js
Comment (nunyvega) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.native.js:35]: Looks like the import for Text is missing

Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.native.js:35]: Oh, good catch 🤦. I'll update it as soon as possible, thanks!

Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.native.js:35]: This made me realize that we haven't covered the upload failure case, which doesn't work as expected. I'll push a commit with a fix and update the testing instructions to cover the case. Sorry for the inconvenience 🙇.

cc @SiobhyB @jhnstn I'll let you know when it's ready in case you could do another review. Thanks!

Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.native.js:35]: Since the block is under development, I thought twice and I'm leaning toward pushing only the fix to import the `Text` component in order to unblock the merge of this PR. The rationale is that handling the cases of failure implies changing different parts and I think it will be easier if we tackle it in a separate PR.

Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.native.js:35]: Fix applied in https://github.com/Automattic/jetpack/pull/29461/commits/4c7a12e8130ba84eb8593fb51b43ca4f7e7b7a81.

Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.native.js:35]: > [..] and update the testing instructions to cover the case. [...]

FYI: I've added the test case "Upload fail" to the **Testing instructions**.

**Review by jhnstn — APPROVED** [review]
Latest updates since my last review look solid, :shipit:

**Comment by fluiddot** [comment]
The PR is already approved so I'm considering merging it. 
- @SiobhyB -  you previously approved it but let me know if you'd like to take another look.
- @nunyvega - I already addressed [the issue you mentioned](https://github.com/Automattic/jetpack/pull/29461#discussion_r1138385814). Let me know if you'd like to take another look.

Thanks!

**Comment by SiobhyB** [comment]
@fluiddot, the changes look good to me as well as your rationale for iterating on the failed upload flow in separate PRs. Thank you for implementing this! 🙇‍♀️

**Pull request merged**
